File: platform/android-seller/app/src/main/java/ca/shawcliffe/seller/DashboardViewModel.kt
package ca.shawcliffe.seller

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.time.Instant
import java.util.Date
import java.util.Locale
import java.util.TimeZone

class DashboardViewModel(private val clientId: String) : ViewModel() {
    var todayStatus by mutableStateOf<DailyStatus?>(null)
        private set
    var products by mutableStateOf<List<Product>>(emptyList())
        private set
    var preorders by mutableStateOf<List<PreorderDetail>>(emptyList())
        private set
    var isLoading by mutableStateOf(false)
        private set
    var isSaving by mutableStateOf(false)
        private set
    var lastSaved by mutableStateOf<String?>(null)
        private set
    var errorMessage by mutableStateOf<String?>(null)

    var isSendingSMS by mutableStateOf(false)
        private set
    var smsResult by mutableStateOf<BroadcastResponse?>(null)
        private set
    var isSendingEmail by mutableStateOf(false)
        private set
    var emailResult by mutableStateOf<BroadcastResponse?>(null)
        private set
    var isSendingPush by mutableStateOf(false)
        private set
    var pushResult by mutableStateOf<BroadcastResponse?>(null)
        private set
    var isSendingTest by mutableStateOf(false)
        private set
    var testResult by mutableStateOf<TestSendResult?>(null)
        private set

    private val today: String = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
        timeZone = TimeZone.getDefault()
    }.format(Date())

    fun clearError() {
        errorMessage = null
    }

    fun loadAll() {
        viewModelScope.launch {
            isLoading = true
            listOf(
                async { loadStatus() },
                async { loadProducts() },
                async { loadPreorders() },
            ).awaitAll()
            isLoading = false
        }
    }

    private suspend fun loadStatus() {
        try {
            // Unique (client_id, date) constraint guarantees 0 or 1 rows.
            todayStatus = supabase.from("daily_status").select {
                filter {
                    eq("client_id", clientId)
                    eq("date", today)
                }
            }.decodeList<DailyStatus>().firstOrNull()
        } catch (e: Exception) {
            errorMessage = "Couldn't load today's status."
        }
    }

    private suspend fun loadProducts() {
        try {
            products = supabase.from("products").select {
                filter { eq("client_id", clientId) }
                order("sort_order", Order.ASCENDING)
            }.decodeList()
        } catch (e: Exception) {
            errorMessage = "Couldn't load products."
        }
    }

    private suspend fun loadPreorders() {
        try {
            val rows = supabase.from("preorders").select {
                filter { eq("client_id", clientId) }
                order("created_at", Order.DESCENDING)
            }.decodeList<Preorder>()

            if (rows.isEmpty()) {
                preorders = emptyList()
                return
            }

            val customerIds = rows.map { it.customerId }.distinct()
            val customers = supabase.from("customers").select(columns = io.github.jan.supabase.postgrest.query.Columns.raw("id, name, phone, email")) {
                filter {
                    eq("client_id", clientId)
                    isIn("id", customerIds)
                }
            }.decodeList<Customer>()
            val customersById = customers.associateBy { it.id }

            val preorderIds = rows.map { it.id }
            val items = supabase.from("preorder_items").select(columns = io.github.jan.supabase.postgrest.query.Columns.raw("preorder_id, product_id, quantity")) {
                filter {
                    eq("client_id", clientId)
                    isIn("preorder_id", preorderIds)
                }
            }.decodeList<PreorderItem>()
            val productsById = products.associateBy { it.id }

            preorders = rows.map { preorder ->
                val lineItems = items
                    .filter { it.preorderId == preorder.id }
                    .map { (productsById[it.productId]?.name ?: "Unknown item") to it.quantity }
                PreorderDetail(
                    preorder = preorder,
                    customer = customersById[preorder.customerId],
                    items = lineItems,
                )
            }
        } catch (e: Exception) {
            errorMessage = "Couldn't load preorders."
        }
    }

    fun setStatus(status: DailyStatusValue) {
        viewModelScope.launch {
            isSaving = true
            val now = Instant.now().toString()
            try {
                if (todayStatus != null) {
                    supabase.from("daily_status").update(DailyStatusUpdate(status.value, now)) {
                        filter {
                            eq("client_id", clientId)
                            eq("date", today)
                        }
                    }
                } else {
                    todayStatus = supabase.from("daily_status")
                        .insert(DailyStatusInsert(clientId, today, status.value)) { select() }
                        .decodeSingle()
                }
                todayStatus = todayStatus?.copy(status = status.value, updatedAt = now)
                lastSaved = SimpleDateFormat("h:mm a", Locale.US).format(Date())
            } catch (e: Exception) {
                errorMessage = "Couldn't update status."
            }
            isSaving = false
        }
    }

    fun setProductStatus(product: Product, status: ProductStatus) {
        val index = products.indexOfFirst { it.id == product.id }
        if (index < 0) return
        val previous = products
        products = products.toMutableList().also { it[index] = it[index].copy(status = status.value) }
        viewModelScope.launch {
            try {
                supabase.from("products").update(ProductStatusUpdate(status.value)) {
                    filter {
                        eq("client_id", clientId)
                        eq("id", product.id)
                    }
                }
            } catch (e: Exception) {
                products = previous
                errorMessage = "Couldn't update product status."
            }
        }
    }

    fun addProduct(name: String, price: Double?) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch {
            try {
                val inserted = supabase.from("products")
                    .insert(ProductInsert(clientId, trimmed, price, ProductStatus.AVAILABLE.value, products.size)) { select() }
                    .decodeSingle<Product>()
                products = products + inserted
            } catch (e: Exception) {
                errorMessage = "Couldn't add product."
            }
        }
    }

    fun deleteProduct(product: Product) {
        val previous = products
        products = products.filter { it.id != product.id }
        viewModelScope.launch {
            try {
                supabase.from("products").delete {
                    filter {
                        eq("client_id", clientId)
                        eq("id", product.id)
                    }
                }
            } catch (e: Exception) {
                products = previous
                errorMessage = "Couldn't delete product."
            }
        }
    }

    fun endOfDay() {
        viewModelScope.launch {
            isSaving = true
            setStatus(DailyStatusValue.CLOSED)
            val previous = products
            products = products.map { it.copy(status = ProductStatus.SOLD_OUT.value) }
            try {
                supabase.from("products").update(ProductStatusUpdate(ProductStatus.SOLD_OUT.value)) {
                    filter { eq("client_id", clientId) }
                }
            } catch (e: Exception) {
                products = previous
                errorMessage = "Couldn't complete end of day."
            }
            isSaving = false
        }
    }

    fun updatePreorderStatus(detail: PreorderDetail, status: String) {
        val index = preorders.indexOfFirst { it.id == detail.id }
        if (index < 0) return
        val previous = preorders
        preorders = preorders.toMutableList().also {
            it[index] = it[index].copy(preorder = it[index].preorder.copy(status = status))
        }
        viewModelScope.launch {
            try {
                supabase.from("preorders").update(PreorderStatusUpdate(status)) {
                    filter {
                        eq("client_id", clientId)
                        eq("id", detail.id)
                    }
                }
            } catch (e: Exception) {
                preorders = previous
                errorMessage = "Couldn't update preorder."
            }
        }
    }

    fun sendSMSBroadcast(message: String) {
        val trimmed = message.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch {
            isSendingSMS = true
            smsResult = null
            try {
                smsResult = BroadcastService.send("api/broadcast/sms", SMSBroadcastRequest(trimmed))
            } catch (e: Exception) {
                errorMessage = e.message
            }
            isSendingSMS = false
        }
    }

    fun sendEmailBroadcast(subject: String, message: String) {
        val trimmedSubject = subject.trim()
        val trimmedMessage = message.trim()
        if (trimmedSubject.isEmpty() || trimmedMessage.isEmpty()) return
        viewModelScope.launch {
            isSendingEmail = true
            emailResult = null
            try {
                emailResult = BroadcastService.send("api/broadcast/email", EmailBroadcastRequest(trimmedSubject, trimmedMessage))
            } catch (e: Exception) {
                errorMessage = e.message
            }
            isSendingEmail = false
        }
    }

    fun sendPushBroadcast(message: String) {
        val trimmed = message.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch {
            isSendingPush = true
            pushResult = null
            try {
                pushResult = BroadcastService.send("api/broadcast/push", SMSBroadcastRequest(trimmed))
            } catch (e: Exception) {
                errorMessage = e.message
            }
            isSendingPush = false
        }
    }

    fun sendTest(channel: String, smsMessage: String, emailSubject: String, emailMessage: String, pushMessage: String) {
        viewModelScope.launch {
            isSendingTest = true
            testResult = null
            try {
                val response: BroadcastResponse = when (channel) {
                    "sms" -> {
                        val message = smsMessage.trim().ifEmpty { "Test message from your dashboard." }
                        BroadcastService.send("api/broadcast/sms", SMSBroadcastRequest(message, test = true))
                    }
                    "email" -> {
                        val subject = emailSubject.trim().ifEmpty { "Test notification" }
                        val message = emailMessage.trim().ifEmpty { "Test message from your dashboard." }
                        BroadcastService.send("api/broadcast/email", EmailBroadcastRequest(subject, message, test = true))
                    }
                    else -> {
                        val message = pushMessage.trim().ifEmpty { "Test push from your dashboard." }
                        BroadcastService.send("api/broadcast/push", SMSBroadcastRequest(message, test = true))
                    }
                }
                testResult = TestSendResult(
                    channel = channel,
                    ok = response.sent > 0,
                    message = if (response.sent > 0) "Test sent — check your phone/inbox." else (response.errors?.firstOrNull() ?: "Test send failed"),
                )
            } catch (e: Exception) {
                testResult = TestSendResult(channel = channel, ok = false, message = e.message ?: "Test send failed")
            }
            isSendingTest = false
        }
    }
}

File: platform/ios-seller/ShawcliffeSeller/DashboardViewModel.swift
import Foundation

@MainActor
final class DashboardViewModel: ObservableObject {
    @Published var todayStatus: DailyStatus?
    @Published var products: [Product] = []
    @Published var preorders: [PreorderDetail] = []
    @Published var isLoading = false
    @Published var isSaving = false
    @Published var lastSaved: String?
    @Published var errorMessage: String?
    @Published var isSendingSMS = false
    @Published var smsResult: BroadcastResponse?
    @Published var isSendingEmail = false
    @Published var emailResult: BroadcastResponse?
    @Published var isSendingPush = false
    @Published var pushResult: BroadcastResponse?
    @Published var isSendingTest = false
    @Published var testResult: TestSendResult?

    private let clientId: String
    private let today: String = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = .current
        return formatter.string(from: Date())
    }()

    init(clientId: String) {
        self.clientId = clientId
    }

    func loadAll() async {
        isLoading = true
        async let statusTask: Void = loadStatus()
        async let productsTask: Void = loadProducts()
        async let preordersTask: Void = loadPreorders()
        _ = await (statusTask, productsTask, preordersTask)
        isLoading = false
    }

    func loadStatus() async {
        do {
            // Unique (client_id, date) constraint guarantees 0 or 1 rows.
            let rows: [DailyStatus] = try await supabase
                .from("daily_status")
                .select()
                .eq("client_id", value: clientId)
                .eq("date", value: today)
                .execute()
                .value
            todayStatus = rows.first
        } catch {
            errorMessage = "Couldn't load today's status."
        }
    }

    func loadProducts() async {
        do {
            let items: [Product] = try await supabase
                .from("products")
                .select()
                .eq("client_id", value: clientId)
                .order("sort_order")
                .execute()
                .value
            products = items
        } catch {
            errorMessage = "Couldn't load products."
        }
    }

    func loadPreorders() async {
        do {
            let rows: [Preorder] = try await supabase
                .from("preorders")
                .select()
                .eq("client_id", value: clientId)
                .order("created_at", ascending: false)
                .execute()
                .value

            guard !rows.isEmpty else {
                preorders = []
                return
            }

            let customerIds = Array(Set(rows.map(\.customerId)))
            let customers: [Customer] = try await supabase
                .from("customers")
                .select("id, name, phone, email")
                .eq("client_id", value: clientId)
                .in("id", values: customerIds)
                .execute()
                .value
            let customersById = Dictionary(uniqueKeysWithValues: customers.map { ($0.id, $0) })

            let preorderIds = rows.map(\.id)
            let items: [PreorderItem] = try await supabase
                .from("preorder_items")
                .select("preorder_id, product_id, quantity")
                .eq("client_id", value: clientId)
                .in("preorder_id", values: preorderIds)
                .execute()
                .value
            let productsById = Dictionary(uniqueKeysWithValues: products.map { ($0.id, $0) })

            preorders = rows.map { preorder in
                let lineItems = items
                    .filter { $0.preorderId == preorder.id }
                    .map { (productsById[$0.productId]?.name ?? "Unknown item", $0.quantity) }
                return PreorderDetail(
                    preorder: preorder,
                    customer: customersById[preorder.customerId],
                    items: lineItems
                )
            }
        } catch {
            errorMessage = "Couldn't load preorders."
        }
    }

    func setStatus(_ status: DailyStatusValue) async {
        isSaving = true
        let now = ISO8601DateFormatter().string(from: Date())
        do {
            if todayStatus != nil {
                try await supabase
                    .from("daily_status")
                    .update(DailyStatusUpdate(status: status.rawValue, updated_at: now))
                    .eq("client_id", value: clientId)
                    .eq("date", value: today)
                    .execute()
            } else {
                let inserted: DailyStatus = try await supabase
                    .from("daily_status")
                    .insert(DailyStatusInsert(client_id: clientId, date: today, status: status.rawValue))
                    .select()
                    .single()
                    .execute()
                    .value
                todayStatus = inserted
            }
            if var current = todayStatus {
                current.status = status
                current.updatedAt = now
                todayStatus = current
            }
            lastSaved = DateFormatter.localizedString(from: Date(), dateStyle: .none, timeStyle: .short)
        } catch {
            errorMessage = "Couldn't update status."
        }
        isSaving = false
    }

    func setProductStatus(_ product: Product, status: ProductStatus) async {
        guard let index = products.firstIndex(where: { $0.id == product.id }) else { return }
        let previous = products[index].status
        products[index].status = status
        do {
            try await supabase
                .from("products")
                .update(ProductStatusUpdate(status: status.rawValue))
                .eq("client_id", value: clientId)
                .eq("id", value: product.id)
                .execute()
        } catch {
            products[index].status = previous
            errorMessage = "Couldn't update product status."
        }
    }

    func addProduct(name: String, price: Double?) async {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        do {
            let inserted: Product = try await supabase
                .from("products")
                .insert(ProductInsert(
                    client_id: clientId,
                    name: trimmed,
                    price: price,
                    status: ProductStatus.available.rawValue,
                    sort_order: products.count
                ))
                .select()
                .single()
                .execute()
                .value
            products.append(inserted)
        } catch {
            errorMessage = "Couldn't add product."
        }
    }

    func deleteProduct(_ product: Product) async {
        let previous = products
        products.removeAll { $0.id == product.id }
        do {
            try await supabase
                .from("products")
                .delete()
                .eq("client_id", value: clientId)
                .eq("id", value: product.id)
                .execute()
        } catch {
            products = previous
            errorMessage = "Couldn't delete product."
        }
    }

    func endOfDay() async {
        isSaving = true
        await setStatus(.closed)
        let previous = products
        products = products.map { var p = $0; p.status = .soldOut; return p }
        do {
            try await supabase
                .from("products")
                .update(ProductStatusUpdate(status: ProductStatus.soldOut.rawValue))
                .eq("client_id", value: clientId)
                .execute()
        } catch {
            products = previous
            errorMessage = "Couldn't complete end of day."
        }
        isSaving = false
    }

    func updatePreorderStatus(_ detail: PreorderDetail, status: String) async {
        guard let index = preorders.firstIndex(where: { $0.id == detail.id }) else { return }
        let previous = preorders[index].preorder.status
        preorders[index].preorder.status = status
        do {
            try await supabase
                .from("preorders")
                .update(PreorderStatusUpdate(status: status))
                .eq("client_id", value: clientId)
                .eq("id", value: detail.id)
                .execute()
        } catch {
            preorders[index].preorder.status = previous
            errorMessage = "Couldn't update preorder."
        }
    }

    func sendSMSBroadcast(message: String) async {
        let trimmed = message.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        isSendingSMS = true
        smsResult = nil
        do {
            smsResult = try await BroadcastService.send(
                path: "api/broadcast/sms",
                body: SMSBroadcastRequest(message: trimmed)
            )
        } catch {
            errorMessage = error.localizedDescription
        }
        isSendingSMS = false
    }

    func sendEmailBroadcast(subject: String, message: String) async {
        let trimmedSubject = subject.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedMessage = message.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedSubject.isEmpty, !trimmedMessage.isEmpty else { return }
        isSendingEmail = true
        emailResult = nil
        do {
            emailResult = try await BroadcastService.send(
                path: "api/broadcast/email",
                body: EmailBroadcastRequest(subject: trimmedSubject, message: trimmedMessage)
            )
        } catch {
            errorMessage = error.localizedDescription
        }
        isSendingEmail = false
    }

    func sendPushBroadcast(message: String) async {
        let trimmed = message.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        isSendingPush = true
        pushResult = nil
        do {
            pushResult = try await BroadcastService.send(
                path: "api/broadcast/push",
                body: SMSBroadcastRequest(message: trimmed)
            )
        } catch {
            errorMessage = error.localizedDescription
        }
        isSendingPush = false
    }

    func sendTest(channel: String, smsMessage: String, emailSubject: String, emailMessage: String, pushMessage: String) async {
        isSendingTest = true
        testResult = nil
        do {
            let response: BroadcastResponse
            switch channel {
            case "sms":
                let message = smsMessage.trimmingCharacters(in: .whitespacesAndNewlines)
                response = try await BroadcastService.send(
                    path: "api/broadcast/sms",
                    body: SMSBroadcastRequest(message: message.isEmpty ? "Test message from your dashboard." : message, test: true)
                )
            case "email":
                let subject = emailSubject.trimmingCharacters(in: .whitespacesAndNewlines)
                let message = emailMessage.trimmingCharacters(in: .whitespacesAndNewlines)
                response = try await BroadcastService.send(
                    path: "api/broadcast/email",
                    body: EmailBroadcastRequest(
                        subject: subject.isEmpty ? "Test notification" : subject,
                        message: message.isEmpty ? "Test message from your dashboard." : message,
                        test: true
                    )
                )
            default:
                let message = pushMessage.trimmingCharacters(in: .whitespacesAndNewlines)
                response = try await BroadcastService.send(
                    path: "api/broadcast/push",
                    body: SMSBroadcastRequest(message: message.isEmpty ? "Test push from your dashboard." : message, test: true)
                )
            }
            testResult = TestSendResult(
                channel: channel,
                ok: response.sent > 0,
                message: response.sent > 0 ? "Test sent — check your phone/inbox." : (response.errors?.first ?? "Test send failed")
            )
        } catch {
            testResult = TestSendResult(channel: channel, ok: false, message: error.localizedDescription)
        }
        isSendingTest = false
    }
}
